include/seqan3/range/container/small_string.hpp

```diff
@@ -0,0 +1,265 @@
+    }
+
+    /*!\brief Assign from literal.
+     * \param _lit The literal to assign the string from.
+     *
+     * ### Complexity
+     *
+     * Linear in the size of _lit.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    template <size_t N>
+    constexpr small_string & operator=(char const (&_lit)[N]) noexcept
+    {
+        static_assert(N <= capacity_ + 1, "Length of string literal exceeds capacity of small_string.");
+        assign(_lit);
+        return *this;
+    }
+
+    /*!\brief Assign from literal.
+     * \param _lit The literal to assign the string from.
+     *
+     * ### Complexity
+     *
+     * Linear in the size of _lit.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    template <size_t N>
+    constexpr void assign(char const (&_lit)[N]) noexcept
+    {
+        static_assert(N <= capacity_ + 1, "Length of string literal exceeds capacity of small_string.");
+        assert(_lit[N - 1] == '\0');
+        clear();
+        assign(&_lit[0], &_lit[N-1]);
+    }
+    //!\}
+
+    /*!\name Capacity
+     * \{
+     */
+    //!\brief Returns the maximal size which equals the capacity.
+    constexpr size_type max_size() const noexcept
+    {
+        return capacity_;
+    }
+
+    //!\brief Returns the maximal capacity.
+    constexpr size_type capacity() const noexcept
+    {
+        return capacity_;
+    }
+    //!\}
+
+    /*!\name Element access
+     * \{
+     */
+    /*!\brief Returns the content represented as std::string.
+     *
+     * \returns `std::string` The stored string.
+     *
+     * ### Exceptions
+     *
+     * Strong exception guarantee. No data is modified.
+     *
+     * ### Complexity
+     *
+     * Linear in the size of the string.
+     */
+    std::string string() const
+    {
+        return std::string{this->cbegin(), this->cend()};
+    }
+
+    /*!\brief Returns the content represented as 0-terminated c-style string.
+     *
+     * \returns `const char *` The stored string.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     *
+     * ### Complexity
+     *
+     * Constant.
+     */
+    constexpr const char * c_str() const noexcept
+    {
+        return this->data_.data();
+    }
+    //!\}
+
+    /*!\name Modifiers
+     * \{
+     */
+    //!\copydoc seqan3::small_vector::clear
+    constexpr void clear() noexcept
+    {
+        *this = small_string{};
+    }
+
+    //!\copydoc seqan3::small_vector::pop_back
+    constexpr void pop_back() noexcept
+    {
+        assert(this->sz > 0);
+        --this->sz;
+        this->data_[this->sz] = '\0';
+    }
+
+    //!\copydoc seqan3::small_vector::resize(size_type const)
+    constexpr void resize(size_type const count) noexcept
+    {
+        assert(count <= capacity_);
+
+        for (size_t i = this->sz; i <= count; ++i)
+            this->data_[i] = '\0';
+
+        this->sz = count;
+        this->data_[this->sz] = '\0';
+    }
+    //!\}
```

Comment:
you are only providing one of the resize() overloads, what about the other one? How does that one ensure that the string end null-terminated? What about `insert()` at the end?

Reply:
What about `push_back()`?

Reply:
> what about the other one? 

I thought since you provide a value with which to fill the remaining entries, this is usually done when resizing to a bigger size (which would not be a problem because the array contains `\0` at every position after its `.size()`). But you are right, in case I resize to a smaller string this is not correct.

> What about insert() at the end? What about push_back()?

Since the array contains `\0` at every position after its `.size()`, it will be automatically zero terminated.
That's why clear() sets all the `\0` characters again.

Reply:
> (which would not be a problem because the array contains \0 at every position after its .size())

This is the core question. Guaranteeing that makes implementation simpler, but is more expensive...

Reply:
not 100% convinced yet :D